feat: Add Hours, Minutes, Seconds fields while taking input as idleTimeLimit

thepiyush-303 opened the pull request:
<!-- This is a pull request template, you do not need to uncomment or remove the comments, they won't show up in the PR text. -->

<!-- Your Pull Request name should start with one of the following tags
  feat: Adding a new feature
  refactor: A code change that doesn't change behavior (it doesn't add anything and doesn't fix anything)
  fix: For bug fixes that affect the end-user
  chore: For small tasks
  docs: For documentation
  ci: For updating CI configuration
  test: For adding tests
  i18n: For updating any translations
  regression: Issues created/reported/fixed during the development phase. kind of problem that never existed in production and that we don't need to list in a changelog for the end user
-->

<!-- Checklist!!! If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code. 
  - I have read the Contributing Guide - https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat doc
  - I have signed the CLA - https://cla-assistant.io/RocketChat/Rocket.Chat
  - Lint and unit tests pass locally with my changes
  - I have added tests that prove my fix is effective or that my feature works (if applicable)
  - I have added necessary documentation (if applicable)
  - Any dependent changes have been merged and published in downstream modules
-->

## Proposed changes (including videos or screenshots)
<!--
  Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
  If it fixes a bug or resolves a feature request, be sure to link to that issue below.
  This description won't be displayed to our end users in the release notes, so feel free to add as much technical context as needed.
  If the changes introduced in this pull request must be presented in the release notes, make sure to add a changeset file. Check our guidelines for adding a changeset to your pull request: https://developer.rocket.chat/contribute-to-rocket.chat/modes-of-contribution/participate-in-rocket.chat-development/development-workflow#4.-adding-changeset-to-your-pull-request 
-->
Before
![image](https://github.com/user-attachments/assets/fac8adfe-a461-4938-bac8-90c3c7bcb3a3)

After
![Screenshot from 2024-10-22 21-18-46](https://github.com/user-attachments/assets/eff17615-0683-4a41-8b22-ff0c8c8eecae)


## Issue(s)
<!-- Link the issues being closed by or related to this PR. For example, you can use #594 if this PR closes issue number 594 -->
#33566


<!-- Mention how you would reproduce the bug if not mentioned on the issue page already. Also mention which screens are going to have the changes if applicable -->


<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

thepiyush-303 commented:
Hello @dougfabris ,
Can you please review my changes.

Gustrb commented:
Hey @thepiyush-303 thanks for your contribution :)
There is no need to ping anyone, your PR will be reviewed as soon as possible.

But taking a closer look at it, it concerns me that we would need yo have 3 redundant settings, I think that if we want to follow the approach of having 3 input fields, we could convert their respective values to seconds and just store that, wdyt?

thepiyush-303 commented:
I understand your concern about having redundant settings. Converting the input values to seconds and storing that sounds like a more efficient approach.

However, I’ve encountered a couple of issues while implementing this. Firstly, when I change the field values and save them, they revert to the default values when I navigate back to edit them again.

Additionally, I'm running into an "invalid params" error when I attempt to save the values. This appears to be due to the extra fields not being recognized by the form.

Gustrb commented:
I am not a frontend dev, but I don't think there is a need to store the value the same way the user put in, for example:

the user inserts 61 minutes, we convert it to 3660 seconds and whenever we load the page we display the value as 1 hour and 1 minute.

it does not sound like a complicated algorithm

Gustrb commented on .changeset/wild-sheep-hug.md:
```diff
@@ -0,0 +1,5 @@
+'@rocket.chat/meteor': patch
+---
+
+Enhanced userPresence options by adding hours and minutes fields alongside seconds for idleTimeLimit, allowing more flexible idle time configuration.
```
This changeset is too technical, the user does not know about the internal idleTimeLimit

Gustrb commented on apps/meteor/client/startup/startup.ts:
```diff
@@ -48,7 +48,14 @@ Meteor.startup(() => {
+
+			const idleTimeLimitS = (getUserPreference(user, 'idleTimeLimitS') as number | null | undefined) || 0;
+
+			const idleTimeLimit = (idleTimeLimitH || 0) * 3600 + (idleTimeLimitM || 0)* 60 + (idleTimeLimitS || 0);
```
why is there '|| 0' here?

Gustrb commented on apps/meteor/server/methods/saveUserPreferences.ts:
```diff
@@ -153,10 +159,13 @@ export const saveUserPreferences = async (settings: Partial<UserPreferences>, us
+	const idleTimeLimit = (settings.idleTimeLimitH || 0) * 3600 + (settings.idleTimeLimitM || 0) * 60 + (settings.idleTimeLimitS || 0);
+
+	if (settings.idleTimeLimit != null && idleTimeLimit < 60) {
+	throw new Meteor.Error('invalid-idle-time-limit-value', 'Idle time limit must be at least 1 minutes.');
```
please indent this code

thepiyush-303 commented:
> I am not a frontend dev, but I don't think there is a need to store the value the same way the user put in, for example:
> 
> the user inserts 61 minutes, we convert it to 3660 seconds and whenever we load the page we display the value as 1 hour and 1 minute.
> 
> it does not sound like a complicated algorithm
> 
> The invalid fields error is probably because the backend is not expecting those new fields you are sending

Thanks for your input! The issue is that the user needs to provide input in hours, minutes, and seconds. Once they enter the values, if they want to see the changes in the fields, the backend would indeed need to support that. That’s why I chose the approach in the PR. If I'm misunderstanding something, please let me know!

thepiyush-303 commented:
Got it, thanks for the clarification! Should I leave this issue as is, or is there anything else you'd like me to address?

Gustrb commented:
I mean, if you have the time I think it would be a cool addition to do it in the client

thepiyush-303 commented:
> I mean, if you have the time I think it would be a cool addition to do it in the client

but since you mentioned that having multiple settings might not be approved by other collaborators, I'm hesitant to continue with this approach. If it’s a significant enhancement, I’d be happy to explore it further if it aligns with the team's direction.

Gustrb commented:
What I mean by that is, we have user preferences that we store in the database, I don't think those inputs should be mapped to actual values in the database, I think it would be interesting to convert their values to seconds and keep storing seconds.
The only change that I think would fit in this case is adding the inputs so we can have a better granularity in that field, I don't think that granularity should go all the way to the databse, it should be nothing more than an abstraction.

thepiyush-303 commented:
@Gustrb 
I have made the requested changes.

thepiyush-303 commented:
i have made a new commit making the required changes 50ab19e64d31d5987b7c8a8f4553afa84dca5ec1

Gustrb commented on .changeset/wild-sheep-hug.md:
```diff
@@ -0,0 +1,5 @@
+'@rocket.chat/meteor': patch
+---
+
+Enhanced userPresence options making it more human readable.
```
Since this is customer facing you could add more details

```
Added input fields so the user can specify exactly the time of the enableAutoAway preference in hours, minutes and seconds.
```

Gustrb commented on apps/meteor/client/startup/startup.ts:
```diff
@@ -48,7 +48,14 @@ Meteor.startup(() => {
 
 		if (getUserPreference(user, 'enableAutoAway')) {
-			const idleTimeLimit = (getUserPreference(user, 'idleTimeLimit') as number | null | undefined) || 300;
+			const idleTimeLimitH = (getUserPreference(user, 'idleTimeLimitH') as number | null | undefined) || 0;
```
I think we can do this without touching this file

Gustrb commented on apps/meteor/client/views/account/preferences/PreferencesUserPresenceSection.tsx:
```diff
@@ -27,14 +30,19 @@ const PreferencesUserPresenceSection = () => {
 				</Field>
 				<Field>
-					<FieldLabel htmlFor={idleTimeLimit}>{t('Idle_Time_Limit')}</FieldLabel>
+					<FieldLabel htmlFor={idleTimeLimitH}>{t('Idle_Time_Limit')}</FieldLabel>
```
don't you have access to a OnChange callback or smth?
so you can update idleTimeLimit without needing to have 3 separate settings

Gustrb commented on apps/meteor/client/views/account/preferences/useAccountPreferencesValues.ts:
```diff
@@ -43,8 +45,10 @@ export const useAccountPreferencesValues = (): AccountPreferencesData => {
 	const enableAutoAway = useUserPreference<boolean>('enableAutoAway');
-	const idleTimeLimit = useUserPreference<number>('idleTimeLimit');
-
+	const idleTimeLimitH = useUserPreference<number>('idleTimeLimitH');
```
we should remove those extra settings leaving only one

Gustrb commented on apps/meteor/server/methods/saveUserPreferences.ts:
```diff
@@ -153,10 +159,13 @@ export const saveUserPreferences = async (settings: Partial<UserPreferences>, us
 
-	if (settings.idleTimeLimit != null && settings.idleTimeLimit < 60) {
-		throw new Meteor.Error('invalid-idle-time-limit-value', 'Invalid idleTimeLimit');
+	const idleTimeLimit = (settings.idleTimeLimitH || 0) * 3600 + (settings.idleTimeLimitM || 0) * 60 + (settings.idleTimeLimitS || 0);
```
this won't have to be changed

thepiyush-303 commented:
As per your suggestion, I'm working on the second approach to resolve this issue, but I'm currently stuck on a particular challenge.

When I modify the seconds, minutes, and hours fields, I'm able to convert and store these values as a total in seconds. However, I'm unsure how to effectively bind this total seconds value to the component which can register total seconds as 'idleTimeLimit' . If you'd like more details, I'm happy to raise a new PR for further clarification.

thepiyush-303 commented:
Hello @Gustrb ,
I have made the changes you suggested. Specifically:

   1. I updated my approach to storing values in three different settings, aligning with your earlier suggestion.
   2. I have tested the changes, and everything is working fine.

Additionally, I have added a new file. Could you please advise on the appropriate location for this file?

Finally, could you suggest any changes needed to ensure that my code adheres to Rocket's coding standards?

thepiyush-303 commented:
How can i solve my `deploy-preview` check?
if i make any mistake while setting up the project locally?

thepiyush-303 commented:
closing this because of bad commits.
will create a new PR
